webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/facts/nodes/NodeFactRepository.scala

```diff
@@ -0,0 +1,241 @@
+
+  def fileFormat: String
+  def entity: String
+
+  def toJson(data: B): IOResult[String]
+
+  // this is just a relative path from a virtual root, for example for node it will be: "accepted/node-uuid.json"
+  def getEntityPath(id: A): String
+
+}
+
+
+/*
+ * We have only one git for all fact repositories. This is the one managing semaphore, init, etc.
+ * All fact repositories will be a subfolder on it:
+ * - /var/rudder/fact-repository/nodes
+ * - /var/rudder/fact-repository/rudder-config
+ * - /var/rudder/fact-repository/groups
+ * etc
+ */
+
+
+class GitNodeFactRepository(
+    override val gitRepo   : GitRepositoryProvider
+  ,              groupOwner: String
+) extends
+  NodeFactRepository with
+  GitItemRepository with
+  SerializeFacts[(NodeId, InventoryStatus),(NodeInfo, FullInventory, Seq[Software])] {
+
+  override val relativePath = "nodes"
+  override val entity: String = "node"
+  override val fileFormat: String = "1"
+  val commiter = new PersonIdent("rudder-fact", "email not set")
+
+  override def getEntityPath(id: (NodeId, InventoryStatus)): String = {
+    s"${id._2.name}/${id._1.value}.json"
+  }
+
+
+  def getFile(id: NodeId, status: InventoryStatus): File = {
+    gitRepo.rootDirectory / relativePath / getEntityPath((id, status))
+  }
+
+  /*
+     * serialize the inventory into a normalized JSON string.
+     * As we want it to be human readable and searchable, we will use an indented format.
+     */
+  def toJson(data: (NodeInfo, FullInventory, Seq[Software])): IOResult[String] = {
+    val (nodeInfo, inventory, software) = data
+    // we want to store objects alphabetically
+    val inv = inventory
+                .modify(_.node.accounts).using(_.sorted)
+                .modify(_.node.customProperties).using(_.sortBy(_.name))
+                .modify(_.node.environmentVariables).using(_.sortBy(_.name))
+                .modify(_.node.fileSystems).using(_.sortBy(_.name))
+                .modify(_.node.networks).using(_.sortBy(_.name))
+                .modify(_.node.processes).using(_.sortBy(_.commandName))
+                .modify(_.machine.each.bios).using(_.sortBy(_.name))
+                .modify(_.machine.each.controllers).using(_.sortBy(_.name))
+                .modify(_.machine.each.memories).using(_.sortBy(_.name))
+                .modify(_.machine.each.ports).using(_.sortBy(_.name))
+                .modify(_.machine.each.processors).using(_.sortBy(_.name))
+                .modify(_.machine.each.slots).using(_.sortBy(_.name))
+                .modify(_.machine.each.sounds).using(_.sortBy(_.name))
+                .modify(_.machine.each.storages).using(_.sortBy(_.name))
+                .modify(_.machine.each.videos).using(_.sortBy(_.name))
+
+
+    val json = FullDetailLevel.toJson(nodeInfo, inventory.node.main.status, None, Some(inv), software.sortBy(_.name.getOrElse("")))
+    // add entity type and file format at the begining
+    ("entity" -> entity ) ~ ( "fileFormat" -> fileFormat ) ~ json
```

Comment:
I think you want renderJson in the trait that adds entity and fileFormat
and a toJson that renders the json   

something in the trait 

``` 
def renderJson(b : B) : IOResult[String] = {
  val json = ("entity" -> entity ) ~ ( "fileFormat" -> fileFormat ) ~ toJson(b)
  IOResult.effect(prettyRender(json))
} ```

Reply:
I wanted to hide the `Json` liftweb type to allow using either that one or zio-json one (which doesn't have an intermediate representation). But I agree that entity and fileformat are generic. I'm not sure how to make it generic, so I stuck with the simple thing for now. I will add comment to explain.